File: GoLang/solution/solution0241.go
package solution

import "strconv"

func DiffWaysToCompute(expression string) []int {
	result := []int{}
	count := 0
	for i := 0; i < len(expression); i++ {
		if expression[i] == '+' || expression[i] == '-' || expression[i] == '*' {
			count++
		}
	}
	if count == 0 {
		num, _ := strconv.Atoi(expression)
		result = append(result, num)
		return result
	}

	for i := 0; i < len(expression); i++ {
		if expression[i] == '+' || expression[i] == '-' || expression[i] == '*' {
			left := DiffWaysToCompute(expression[:i])
			right := DiffWaysToCompute(expression[i+1:])
			for _, l := range left {
				for _, r := range right {
					switch expression[i] {
					case '+':
						result = append(result, l + r)
					case '-':
						result = append(result, l - r)
					case '*':
						result = append(result, l * r)
					}
				}
			}
		}
	}
	return result
}

File: JavaR2/src/Solution0241.java
import java.util.ArrayList;
import java.util.List;

/**
 * example: 2-1-1
 * result: 2, 0
 */
public class Solution0241 {
	public List<Integer> diffWaysToCompute(String expression) {
		List<Integer> result = new ArrayList<>();
		if (!expression.contains("+") && !expression.contains("-") && !expression.contains("*")) {
			result.add(Integer.parseInt(expression));
			return result;
		}

		for (int i = 0; i < expression.length(); i++) {
			if (expression.charAt(i) == '+' || expression.charAt(i) == '-' || expression.charAt(i) == '*') {
				List<Integer> leftList = diffWaysToCompute(expression.substring(0, i));
				List<Integer> rightList = diffWaysToCompute(expression.substring(i+1));
				for (Integer left : leftList) {
					for (Integer right : rightList) {
						if (expression.charAt(i) == '+') {
							result.add(left+right);
						} else if (expression.charAt(i) == '-') {
							result.add(left-right);
						} else if (expression.charAt(i) == '*') {
							result.add(left*right);
						}
					}
				}
			}
		}

		return result;
	}
}
